### PR TITLE
restore: fix container restore into pod

### PR DESCRIPTION
**Pull request by rst0git**
Currently, when Podman restores a container into a Pod, it always fails with the following error:

    Error: cannot add container f96670b26e53e70f7f451191ea39a093c940c6c48b47218aeeef1396cb860042 to pod h2-pod: no such pod

#### Steps to reproduce this error:

```console
podman pod create --name=test
podman run -d --pod=test --name looper busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman container checkpoint -l --export=/tmp/test.tar.gz
podman rm -a
podman pod rm -a

podman pod create --name=test
podman container restore --pod=test --import=/tmp/test.tar.gz
```

This error occurs because [`r.state.Pod()`](https://github.com/containers/podman/blob/b8d95a5893572b37c8257407e964ad06ba87ade6/libpod/boltdb_state.go#L2087) is called in [`setupContainer()`](https://github.com/containers/podman/blob/b8d95a5893572b37c8257407e964ad06ba87ade6/libpod/runtime_ctr.go#L354) with the Pod *name* instead of *ID*. This problem is fixed by setting `ctrConfig.Pod` to `pod.ID()`.

#### Does this PR introduce a user-facing change?
None

```release-note
Fixed container restore into Pod.
```

### CODE REVIEW

**Comment by Luap99** [comment]
Please make sure to add a test

**Review by stano45 — APPROVED** [review]
I tested locally, the fix works. Thanks a lot!

File: test/e2e/checkpoint_test.go [hunk 1]
```diff
@@ -1110,6 +1110,87 @@ var _ = Describe("Podman checkpoint", func() {
 		"net,uts",
 		"uts,pid",
 	}
+
+	It("podman checkpoint and restore container into Pod", func() {
```
Comment (Luap99) [inline test/e2e/checkpoint_test.go:1114]: Can you just add the podName to one of the existing pod restore tests? These tests are very slow so we should combine many things if possible.

Comment (rst0git) [reply test/e2e/checkpoint_test.go]: Thank you @Luap99! I've updated the pull request to run half of the existing pod restore tests with Pod name and other half with Pod ID.

**Comment by mheon** [comment]
LGTM

File: test/e2e/checkpoint_test.go [hunk 2]
```diff
@@ -1218,25 +1213,18 @@ var _ = Describe("Podman checkpoint", func() {
...
-				"-i",
-				fileName,
-			})
+			// Restore container into Pod.
+			// Verify that restore works with both Pod name and ID.
+			if index%2 == 1 {
+				result = podmanTest.Podman([]string{"container", "restore", "--pod", podName, "-i", fileName})
+			} else {
+				result = podmanTest.Podman([]string{"container", "restore", "--pod", podID, "-i", fileName})
+			}
```
Comment (edsantiago) [inline test/e2e/checkpoint_test.go:1227]: This is 80% duplication. Would you consider cleaning up so the variable part is the only difference?
```go
    podArg := podName
    if index%2 == 1 {
        podArg = podID
    }
    result = ... podArg,
```

**Review by edsantiago — CHANGES_REQUESTED** [review]
Tests are being skipped:
```
[SKIPPED] runtime does not support pod restore: crun
```
Relevant code: https://github.com/containers/podman/blob/298f31ba6fc1efcf6630282c45bb80b3f95f2534/pkg/checkpoint/crutils/checkpoint_restore_utils.go#L225-L233

Is there some CI host on which `crun --lsm-mount-context`  is implemented? I would really like to see a successful run, *and* a failing run of these tests when run against main.

One nit. I am blocking because of no-tests, not because of my duplication nit.

**Comment by edsantiago** [comment]
Ran ginkgo tests on my laptop using `--runc` and see consistent failures of the form "expected 1 to equal 2",  
https://github.com/containers/podman/blob/298f31ba6fc1efcf6630282c45bb80b3f95f2534/test/e2e/checkpoint_test.go#L1243

**Comment by rst0git** [comment]
> Is there some CI host on which crun restore --lsm-mount-context is implemented?

The `--lsm-mount-context` option was implemented only in runc (https://github.com/opencontainers/runc/pull/3068). crun currently doesn't support this option. When crun is used, Podman  would show the following error message:
```
$ sudo podman container restore --pod=test --import=/tmp/test.tar.gz
Error: runtime /usr/bin/crun does not support pod restore
```

File: test/e2e/checkpoint_test.go (hunk 2 above)
Comment (rst0git) [reply test/e2e/checkpoint_test.go]: @edsantiago Thank you for the suggestion! I've updated the pull request with this change.

**Comment by rst0git** [comment]
> Ran ginkgo tests on my laptop using --runc and see consistent failures of the form "expected 1 to equal 2",

When running the tests locally as shown below, they pass
```console
$ git diff
diff --git a/test/e2e/checkpoint_test.go b/test/e2e/checkpoint_test.go
index 58198fe37..de571139a 100644
--- a/test/e2e/checkpoint_test.go
+++ b/test/e2e/checkpoint_test.go
@@ -1119,7 +1119,7 @@ var _ = Describe("Podman checkpoint", func() {
                share := share // copy into local scope, for use inside function
                index := index
 
-               It(testName, func() {
+               FIt(testName, func() {
                        podName := "test_pod"
 
                        if err := criu.CheckForCriu(criu.PodCriuVersion); err != nil {

$ sudo make localintegration FOCUS_FILE=checkpoint_test.go OCI_RUNTIME=runc
...
  Timeline >>
  > Enter [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 06:28:58.155
  < Exit [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 06:28:58.155 (0s)
  > Enter [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 06:28:58.156
  integration timing results
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net)		64.774286
  Podman checkpoint podman checkpoint and restore container out of and into pod (net,uts,pid)		64.943064
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net,uts,pid)		65.902908
  Podman checkpoint podman checkpoint and restore container out of and into pod (net,uts)		65.931323
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net,uts)		65.936874
  Podman checkpoint podman checkpoint and restore container out of and into pod (uts,pid)		69.145405
  < Exit [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 06:29:00.734 (2.578s)
  << Timeline
------------------------------

Ran 6 of 2226 Specs in 127.597 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 2220 Skipped

```

@edsantiago Would you be able to replicate this problem manually? Do you see any error messages?
```
podman pod create --name=test
podman run -d --pod=test --name looper busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman container checkpoint -l --export=/tmp/test.tar.gz
podman rm -a
podman pod rm -a

podman pod create --name=test
podman container restore --pod=test --import=/tmp/test.tar.gz
```

**Comment by Luap99** [comment]
> > Is there some CI host on which crun restore --lsm-mount-context is implemented?
> 
> The `--lsm-mount-context` option was implemented only in runc ([opencontainers/runc#3068](https://github.com/opencontainers/runc/pull/3068)). crun currently doesn't support this option. When crun is used, Podman would show the following error message:
> 
> ```
> $ sudo podman container restore --pod=test --import=/tmp/test.tar.gz
> Error: runtime /usr/bin/crun does not support pod restore
> ```

Note we dropped all runc testing from CI recently and only test with crun. Is there any chance we can add support to crun as this is what our default is in fedora and other distros?
cc @giuseppe

**Review by Luap99 — APPROVED** [review]
LGTM,
I run the tests locally and they failed on main with the reported error and pass with this patch. The crun problem is unfortunate because it is not tested in CI but not your problem and this fix is clearly correct so we should merge it even without CI support.

**Comment by giuseppe** [comment]
@adrianreber is ` --lsm-mount-context` something we could also have in crun?

**Review by giuseppe — APPROVED** [review]
/lgtm

**Comment by rst0git** [comment]
> Is there any chance we can add support to crun as this is what our default is in fedora and other distros?

Yes, we discussed this with @adrianreber yesterday. I will try to implement support for `--lsm-mount-context` with crun and open a pull request.

**Comment by edsantiago** [comment]
> I run the tests locally and they failed on main with the reported error and pass with this patch. 

Can you share your test command? This is mine, and it fails 100%:
```console
$ sudo ./ginkgo --runc "checkpoint and restore container out of and into pod" 
```
(where `./ginkgo` is a simple homebrewed wrapper). I also tried
```console
$ sudo env FOCUS="checkpoint and restore container out of and into pod" OCI_RUNTIME=runc make localintegration
```
All tests fail with "1 is not 2". The relevant error seems to be:
```
Running: podman [...] restore --pod XXX -i foo.tar.gz                                       
  Error: pod XXX does not share the SOMETHING namespace
```
runc-1.1.12-3.fc40.x86_64  criu-3.19-4.fc40.x86_64

(I know this is merged and this is runc so we probably don't care, but I'm annoying that way)

Attachment: [ginkgo script](https://github.com/user-attachments/files/15927691/ginkgo.txt)

**Comment by rst0git** [comment]
@edsantiago you might need to use `FOCUS=checkpoint_test.go` because these tests are generated with a [for loop](https://github.com/containers/podman/blob/79f0f7707ac1eb639cd9f6719ab650edaf7c543f/test/e2e/checkpoint_test.go#L1113).

**Comment by edsantiago** [comment]
`FOCUS_FILE` maybe? But yes, I also tried `./ginkgo --runc checkpoint_test.go`. Same failures. 

It worries me a little that `podman container restore` can barf with "Error blah blah namespace" while exiting 0.

**Comment by rst0git** [comment]
@edsantiago Would you be able to run the tests in a different environment?

The following is the output of the tests on Fedora 39 with runc v1.1.12 and criu v3.19
```
$ sudo make localintegration FOCUS="checkpoint and restore container out of and into pod" OCI_RUNTIME=runc
...
  Timeline >>
  > Enter [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 13:48:05.147
  < Exit [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 13:48:05.147 (0s)
  > Enter [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 13:48:05.148
  integration timing results
  Podman checkpoint podman checkpoint and restore container out of and into pod (net,uts,pid)		67.081255
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net,uts,pid)		67.084638
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net)		67.194082
  Podman checkpoint podman checkpoint and restore container out of and into pod (ipc,net,uts)		68.621705
  Podman checkpoint podman checkpoint and restore container out of and into pod (net,uts)		69.706094
  Podman checkpoint podman checkpoint and restore container out of and into pod (uts,pid)		72.526000
  < Exit [SynchronizedAfterSuite] TOP-LEVEL - /home/radostin/go/src/github.com/containers/podman/test/e2e/common_test.go:200 @ 06/21/24 13:48:07.555 (2.407s)
  << Timeline
------------------------------

Ran 6 of 2226 Specs in 108.399 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 2220 Skipped


Ginkgo ran 1 suite in 2m0.206898675s
Test Suite Passed
```

**Comment by edsantiago** [comment]
Fails on [f39](https://github.com/user-attachments/files/15928559/20240621-restore-2not1-f39.log) and [f40](https://github.com/user-attachments/files/15928565/20240621-restore-2not1-f40.log) (fresh 1minutetip VMs).

**Comment by rst0git** [comment]
> Fails on [f39](https://github.com/user-attachments/files/15928559/20240621-restore-2not1-f39.log) and [f40](https://github.com/user-attachments/files/15928565/20240621-restore-2not1-f40.log) (fresh 1minutetip VMs).

It looks like the restore command completes successfully, and restores a container with ID `01686bf7`.
```
  Running: /root/go/podman/bin/podman --storage-opt overlay.imagestore=/tmp/podman-e2e-477465866/imagecachedir --root /tmp/podman-e2e-477465866/subtest-127715438/root --runroot /tmp/podman-e2e-477465866/subtest-127715438/runroot --runtime runc --conmon /usr/bin/conmon --network-config-dir /etc/containers/networks --network-backend netavark --cgroup-manager systemd --tmpdir /tmp/podman-e2e-477465866/subtest-127715438 --events-backend file --db-backend sqlite --storage-driver overlay container restore --pod test_pod -i /tmp/podman-e2e-477465866/subtest-127715438/checkpoint-01686bf717523235de4c1e1db90d10711deda62d4682f0fc0fdeeac2fbc32844.tar.gz
  01686bf717523235de4c1e1db90d10711deda62d4682f0fc0fdeeac2fbc32844
```
We expect `NumberOfContainersRunning()` to be 2 because we are supposed to have restored container +  Pod infra container running.

```
  Running: /root/go/podman/bin/podman --root /tmp/podman-e2e-477465866/subtest-127715438/root --runroot /tmp/podman-e2e-477465866/subtest-127715438/runroot --runtime runc --conmon /usr/bin/conmon --network-config-dir /etc/containers/networks --network-backend netavark --cgroup-manager systemd --tmpdir /tmp/podman-e2e-477465866/subtest-127715438 --events-backend file --db-backend sqlite --storage-driver overlay ps -q
  60873afce780
  [FAILED] Expected
      <int>: 1
  to equal
      <int>: 2
  In [It] at: /root/go/podman/test/e2e/checkpoint_test.go:1231 @ 06/21/24 09:29:55.276
```
@edsantiago Would you be able to check what containers are running, and what is the state of the restored container?

**Comment by edsantiago** [comment]
My apologies: "Error blah blah namespace" is a red herring; part of the test.

Here's a reproducer on f40:
```console
# bin/podman --runtime runc pod create --name test_pod --share uts,pid                                                             
e1752af0a7161c998e350dab58e781f8ba5dc69c83b531309b46e31cb293796f                                                                                                
# bin/podman --runtime runc run -d --rm --pod e1752 quay.io/libpod/testimage:20240123 top                                          
Trying to pull quay.io/libpod/testimage:20240123...                                                                                                             
...
f6802669af74233b115780fe9979702ee76a4adf86c87072f4d771929c082d04                                                                                                
# bin/podman --runtime runc container checkpoint -e /tmp/foo.tar.gz f6802                                                          
WARN[0033] freezer not supported: openat2 /sys/fs/cgroup/machine.slice/machine-libpod_pod_e1752af0a7161c998e350dab58e781f8ba5dc69c83b531309b46e31cb293796f.slice
/libpod-f6802669af74233b115780fe9979702ee76a4adf86c87072f4d771929c082d04.scope/cgroup.freeze: no such file or directory                                         
WARN[0033] lstat /sys/fs/cgroup/machine.slice/machine-libpod_pod_e1752af0a7161c998e350dab58e781f8ba5dc69c83b531309b46e31cb293796f.slice/libpod-f6802669af74233b1
15780fe9979702ee76a4adf86c87072f4d771929c082d04.scope: no such file or directory                                                                                
f6802                                                                                                                                                           

# bin/podman --runtime runc ps                                                                                                     
CONTAINER ID  IMAGE                                        COMMAND     CREATED        STATUS             PORTS       NAMES                                      
2225b8721eff  localhost/podman-pause:5.2.0-dev-1718984608              2 minutes ago  Up About a minute              e1752af0a716-infra                         

# bin/podman --runtime runc pod rm e1752                                                                                           
e1752af0a7161c998e350dab58e781f8ba5dc69c83b531309b46e31cb293796f                                                                                                
# bin/podman --runtime runc pod create --name test_pod --share uts,pid                                                             
9e9a447c71a4dfb79da206d4fef0476f54a193129ac82ae2b605f352504ef5c4                                                                                                
# bin/podman --runtime runc container restore --pod 9e9a -i /tmp/foo.tar.gz                                                        
f6802669af74233b115780fe9979702ee76a4adf86c87072f4d771929c082d04                                                                                                

# bin/podman --runtime runc ps                                                                                                     
CONTAINER ID  IMAGE                                        COMMAND     CREATED         STATUS         PORTS       NAMES                                         
544fb7f47829  localhost/podman-pause:5.2.0-dev-1718984608              36 seconds ago  Up 12 seconds              9e9a447c71a4-infra                            
```
Repeating, and removing the `--rm` so I can run logs, I get:
```console
# bin/podman --runtime runc logs e461 | cat -vET
...
^Mtop: can't open '/proc': Permission denied
```

**Pull request merged**